src/apps/basicnat/basicnat.lua

```diff
@@ -0,0 +1,172 @@
+module(..., package.seeall)
+
+bit = require("bit")
+ffi = require("ffi")
+
+--- ### `basicnat` app: Implement http://www.ietf.org/rfc/rfc1631.txt Basic NAT
+
+BasicNAT = {}
+
+function BasicNAT:new ()
+   return setmetatable({index = 1, packets = {}},
+                       {__index=BasicNAT})
+end
+
+local function bytes_to_uint32(a, b, c, d)
+   return a * 2^24 + b * 2^16 + c * 2^8 + d
+end
+
+local ipv4_base = 14 -- Ethernet encapsulated ipv4
+local transport_base = 34 -- tranport layer (TCP/UDP/etc) header start
+local proto_tcp = 6
+local proto_udp = 17
+
+local function uint32_to_bytes(u)
+   a = bit.rshift(bit.band(u, 0xff000000) % 2^32, 24)
+   b = bit.rshift(bit.band(u, 0x00ff0000), 16)
+   c = bit.rshift(bit.band(u, 0x0000ff00), 8)
+   d = bit.band(u, 0x000000ff)
+   return a, b, c, d
```

Comment:
Since 0xff == 0x000000ff, it's possible to write the same function as:

```lua
local function uint32_to_bytes(u)
   local a = bit.rshift(u, 24)
   local b = bit.band(bit.rshift(u, 16), 0xff)
   local c = bit.band(bit.rshift(u, 8), 0xff)
   local d = bit.band(u, 0xff)
   return a, b, c, d
end
```